pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,351 @@
+	reader := bodyReader{
+		r: r,
+	}
+	if err := userBkt.Upload(ctx, dst, reader); err != nil {
+		level.Error(logger).Log("msg", "failed uploading block file to bucket",
+			"destination", dst, "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	level.Debug(logger).Log("msg", "finished uploading block file to bucket",
+		"path", pth)
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func decodeMeta(r io.Reader, name string) (metadata.Meta, error) {
+	dec := json.NewDecoder(r)
+	var meta metadata.Meta
+	if err := dec.Decode(&meta); err != nil {
+		return meta, errors.Wrap(err, fmt.Sprintf("failed decoding %s", name))
+	}
+
+	return meta, nil
+}
+
+func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "received request to complete block upload", "content_length", r.ContentLength)
+
+	rdr, err := userBkt.Get(ctx, path.Join(blockID.String(), tmpMetaFilename))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to download %s from object storage", tmpMetaFilename))
+	}
+	meta, err := decodeMeta(rdr, tmpMetaFilename)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+
+	// Upload meta file so block is considered complete
+	if err := c.uploadMeta(ctx, logger, meta, blockID, block.MetaFilename, userBkt); err != nil {
+		return err
+	}
+
+	if err := userBkt.Delete(ctx, path.Join(blockID.String(), tmpMetaFilename)); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to delete %s from block in object storage", tmpMetaFilename))
+	}
+
+	level.Debug(logger).Log("msg", "successfully completed block upload")
+	return nil
+}
+
+func (c *MultitenantCompactor) sanitizeMeta(logger log.Logger, tenantID string, blockID ulid.ULID,
+	meta *metadata.Meta) error {
+	if meta.Thanos.Labels == nil {
+		meta.Thanos.Labels = map[string]string{}
+	}
+
+	meta.ULID = blockID
+
+	var rejLbls []string
+	for l, v := range meta.Thanos.Labels {
+		switch l {
+		// Preserve these labels
+		case mimir_tsdb.CompactorShardIDExternalLabel:
+		// Remove unused labels
+		case mimir_tsdb.DeprecatedTenantIDExternalLabel, mimir_tsdb.DeprecatedIngesterIDExternalLabel, mimir_tsdb.DeprecatedShardIDExternalLabel:
+			level.Debug(logger).Log("msg", fmt.Sprintf("removing unused external label from %s", block.MetaFilename),
+				"label", l, "value", v)
+			delete(meta.Thanos.Labels, l)
+		default:
+			rejLbls = append(rejLbls, l)
+		}
+	}
+
+	if len(rejLbls) > 0 {
+		level.Warn(logger).Log("msg", fmt.Sprintf("rejecting unsupported external label(s) in %s", block.MetaFilename),
+			"labels", strings.Join(rejLbls, ","))
+		return httpError{
+			message:    fmt.Sprintf("unsupported external label(s): %s", strings.Join(rejLbls, ",")),
+			statusCode: http.StatusBadRequest,
+		}
+	}
```

Comment:
Perhaps we can simplify the code, and return error on first such label. (Message could be just: "external labels not supported")

Reply:
@pstibrany thanks, I've tried to revise accordingly.